### ISSUE TITLE
🐛 Unterminated `<<` followed by `>>>` misparses as generic type arguments; `semicolons: "asNeeded"` output is unparseable

### ISSUE THREAD

**Issue by AlbinoGeek** [biomejs/biome#11880]
### Environment information

<details>

```bash
CLI:
  Version:                      2.5.14

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:           npm
  JS_RUNTIME_NAME:              bun
```

</details>

### Rule name

Parser, not a rule. Cascades into false `correctness/noUndeclaredVariables` and
`correctness/noUnusedVariables`.

### Playground link

https://github.com/AlbinoGeek/biome-asi-shift-misparse

```ts
export function netmask(bits: number): number {
  const mask = 1 << bits
  const m = mask >>> 0
  return m
}
```

```
src/shift.ts:3:3  parse  × expected `,` but instead found `const`
src/shift.ts:3:19 parse  × Expected a parenthesis '(' but instead found '>'.
```

`tsc --noEmit --strict` accepts the same file (exit 0).

Both conditions are required: a statement ending in an unterminated `<<` expression,
and a later `>>>` token.

| Source | Parse errors |
| --- | --- |
| `const mask = 1 << bits` then `const m = mask >>> 0` | **2** |
| the same two lines with semicolons | 0 |
| `const mask = 1 << bits` then `const m = mask + 1` | 0 |
| `const mask = 1 < bits` then `const m = 5 >>> 0` | 0 |

The parser appears to treat `<<` as opening two generic type-argument lists and consume
the later `>>>` as their closing brackets, desyncing it for the rest of the function and
making every subsequent identifier look undeclared.

This matters beyond hand-written style: the file above is exactly what Biome's own
formatter emits under `javascript.formatter.semicolons: "asNeeded"`, so `biome format`
produces source that `biome lint` then cannot parse.

### Expected result

The file parses, as it does with semicolons. `<<` in expression position should not be
speculatively parsed as the start of a generic type-argument list.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct
